File: backend/app/routers/words.py
from fastapi import APIRouter, Depends, Query, UploadFile, File, Form, HTTPException
from sqlalchemy.orm import Session
from typing import Optional
from ..database import get_db
from ..schemas.word import WordResponse, WordStudyRequest, StudyPlanRequest, StudySessionData
from ..services.word_service import word_service
from ..routers.auth import get_current_user
from ..models.user import User

router = APIRouter(prefix="/api/words", tags=["words"])


@router.get("/stats")
async def get_word_stats(
    category: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return word_service.get_word_stats(db, current_user.id, category)


@router.get("/today")
async def get_today_words(
    count: int = Query(20, ge=5, le=100),
    category: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    result = word_service.get_today_words(db, current_user.id, count, category)
    return {
        "review": [
            {
                "id": word.id,
                "word": word.word,
                "phonetic": word.phonetic,
                "meaning": word.meaning,
                "example_sentence": word.example_sentence,
                "difficulty": word.difficulty,
                "frequency": word.frequency,
                "exam_requirement": word.exam_requirement,
                "category": word.category,
                "type": "review"
            }
            for word in result["review"]
        ],
        "new": [
            {
                "id": word.id,
                "word": word.word,
                "phonetic": word.phonetic,
                "meaning": word.meaning,
                "example_sentence": word.example_sentence,
                "difficulty": word.difficulty,
                "frequency": word.frequency,
                "exam_requirement": word.exam_requirement,
                "category": word.category,
                "type": "new"
            }
            for word in result["new"]
        ],
        "review_count": result["review_count"],
        "new_count": len(result["new"]),
        "total_today": result["total_today"]
    }


@router.get("/daily-review")
async def get_daily_review_words(
    category: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    words = word_service.get_daily_review_words(db, current_user.id, category)
    return {
        "words": [
            {
                "id": word.id,
                "word": word.word,
                "phonetic": word.phonetic,
                "meaning": word.meaning,
                "example_sentence": word.example_sentence,
                "exam_requirement": word.exam_requirement,
                "category": word.category,
                "type": "review"
            }
            for word in words
        ],
        "count": len(words)
    }


@router.get("/review-words")
async def get_review_words_by_range(
    time_range: str = Query("today", description="复习时间范围: today(今日), day(近一日), week(近一周), month(近一月), recommended(系统推荐)"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    words = word_service.get_review_words_by_range(db, current_user.id, time_range)
    return {
        "words": [
            {
                "id": word.id,
                "word": word.word,
                "phonetic": word.phonetic,
                "meaning": word.meaning,
                "example_sentence": word.example_sentence,
                "exam_requirement": word.exam_requirement,
                "category": word.category,
                "type": "review"
            }
            for word in words
        ],
        "count": len(words)
    }


@router.get("/categories")
async def get_word_categories(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return {"categories": word_service.get_word_categories(db, current_user.id)}


@router.get("/review")
async def get_review_words(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    items = word_service.get_review_words(db, current_user.id)
    return [
        {
            "id": item.id,
            "word_id": item.word_id,
            "word": item.word.word if hasattr(item, 'word') else "",
            "mastery_level": item.mastery_level,
            "next_review_date": item.next_review_date.isoformat() if item.next_review_date else None,
            "review_count": item.review_count,
            "correct_count": item.correct_count,
            "last_study_date": item.last_study_date.isoformat() if item.last_study_date else None,
            "first_study_date": item.first_study_date.isoformat() if item.first_study_date else None,
            "last_rating": item.last_rating,
            "srs_stage": item.srs_stage
        }
        for item in items
    ]


@router.get("/list")
async def get_word_list(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=10, le=100),
    mastery_level: Optional[str] = Query(None),
    keyword: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    result = word_service.get_word_list(db, current_user.id, page, page_size, mastery_level, keyword, category)

    items = [
        {
            "id": item["id"],
            "word_id": item["word_id"],
            "word": item["word"],
            "phonetic": item["phonetic"],
            "meaning": item["meaning"],
            "example_sentence": item["example_sentence"],
            "mastery_level": item["mastery_level"],
            "next_review_date": item["next_review_date"].isoformat() if item["next_review_date"] else None,
            "review_count": item["review_count"],
            "correct_count": item["correct_count"],
            "last_study_date": item["last_study_date"].isoformat() if item["last_study_date"] else None,
            "first_study_date": item["first_study_date"].isoformat() if item["first_study_date"] else None,
            "last_rating": item["last_rating"],
            "srs_stage": item["srs_stage"]
        }
        for item in result["items"]
    ]

    return {"total": result["total"], "items": items}


@router.post("/study")
async def study_word(
    data: WordStudyRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    result = word_service.study_word(db, current_user.id, data)
    return {
        "id": result.id,
        "mastery_level": result.mastery_level,
        "next_review_date": result.next_review_date.isoformat() if result.next_review_date else None,
        "review_count": result.review_count,
        "correct_count": result.correct_count,
        "last_rating": result.last_rating,
        "srs_stage": result.srs_stage
    }


@router.get("/plan")
async def get_study_plan(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return word_service.get_study_plan(db, current_user.id)


@router.post("/plan")
async def save_study_plan(
    data: StudyPlanRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return word_service.save_study_plan(db, current_user.id, data)


@router.get("/session")
async def get_study_session(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    session = word_service.get_study_session(db, current_user.id)
    return session or {}


@router.post("/session")
async def save_study_session(
    data: dict,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return word_service.save_study_session(db, current_user.id, data)


@router.delete("/session")
async def clear_study_session(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    word_service.clear_study_session(db, current_user.id)
    return {"success": True}


@router.post("/upload-wordbook")
async def upload_wordbook(
    file: UploadFile = File(...),
    category: str = Form("我的词书"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    if not file.filename:
        raise HTTPException(status_code=400, detail="未选择文件")

    if file.size and file.size > 50 * 1024 * 1024:
        raise HTTPException(status_code=400, detail="文件大小超过限制（最大50MB）")

    try:
        contents = await file.read()
        result = word_service.upload_wordbook(db, current_user.id, contents, file.filename, category)
        if not result.get("success"):
            raise HTTPException(status_code=400, detail=result.get("message", "词书导入失败"))
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"词书上传失败: {str(e)}")

File: frontend/src/api/words.ts
import axios from './axios'

export interface Word {
  id: number
  word: string
  phonetic: string | null
  meaning: string
  example_sentence: string | null
  difficulty: number
  frequency: number
  exam_requirement: string
  category?: string
  type?: string
}

export interface UserWord {
  id: number
  word_id: number
  word: string
  phonetic?: string
  meaning?: string
  example_sentence?: string
  mastery_level: string
  next_review_date: string | null
  review_count: number
  correct_count: number
  last_study_date: string | null
  first_study_date: string | null
  last_rating: string | null
  srs_stage: number
}

export interface WordStats {
  total: number
  studied: number
  mastered: number
  today: number
  has_wordbook?: boolean
  review_due?: number
}

export interface StudyPlan {
  daily_word_count: number
  word_category: string | null
  batch_size: number
  study_mode: string
}

export interface TodayWordsResponse {
  review: Word[]
  new: Word[]
  review_count: number
  new_count: number
  total_today: number
}

export interface StudySession {
  current_round: number
  total_rounds: number
  round_queue: RoundQueueItem[]
  round_stats: { known: number; vague: number; unknown: number }
  global_index: number
  total_words_today: number
  study_mode: string
  batch_size: number
  all_word_ids: number[]
  completed_rounds: number
  category: string | null
}

export interface RoundQueueItem {
  wordId: number
  word: string
  phonetic: string | null
  meaning: string
  example_sentence: string | null
  exam_requirement: string
  type: string
  repeatCount: number
}

export async function getWordStats(category?: string): Promise<WordStats> {
  return await axios.get('/words/stats', { params: category ? { category } : {} })
}

export async function getTodayWords(count: number = 20, category?: string): Promise<TodayWordsResponse> {
  return await axios.get('/words/today', { params: { count, category } })
}

export async function getDailyReviewWords(category?: string): Promise<{ words: Word[]; count: number }> {
  return await axios.get('/words/daily-review', { params: { category } })
}

export async function getReviewWordsByRange(timeRange: string = 'today'): Promise<{ words: Word[]; count: number }> {
  return await axios.get('/words/review-words', { params: { time_range: timeRange } })
}

export async function getWordCategories(): Promise<{ categories: string[] }> {
  return await axios.get('/words/categories')
}

export async function getReviewWords(): Promise<UserWord[]> {
  return await axios.get('/words/review')
}

export async function getWordList(params?: {
  page?: number
  page_size?: number
  mastery_level?: string
  keyword?: string
  category?: string
}): Promise<{ total: number; items: UserWord[] }> {
  return await axios.get('/words/list', { params })
}

export async function studyWord(wordId: number, result: string): Promise<UserWord> {
  return await axios.post('/words/study', { word_id: wordId, result })
}

export async function getStudyPlan(): Promise<StudyPlan> {
  return await axios.get('/words/plan')
}

export async function saveStudyPlan(
  dailyWordCount: number,
  wordCategory?: string,
  batchSize?: number,
  studyMode?: string
): Promise<StudyPlan> {
  return await axios.post('/words/plan', {
    daily_word_count: dailyWordCount,
    word_category: wordCategory,
    batch_size: batchSize ?? 20,
    study_mode: studyMode ?? 'mixed'
  })
}

export async function getStudySession(): Promise<StudySession | null> {
  const res: any = await axios.get('/words/session')
  return res && Object.keys(res).length > 0 ? (res as StudySession) : null
}

export async function saveStudySession(session: StudySession): Promise<StudySession> {
  return await axios.post('/words/session', session)
}

export async function clearStudySession(): Promise<{ success: boolean }> {
  return await axios.delete('/words/session')
}

export interface UploadWordbookResult {
  success: boolean
  message: string
  imported_count?: number
  category?: string
}

export async function uploadWordbook(file: File, category?: string): Promise<UploadWordbookResult> {
  const formData = new FormData()
  formData.append('file', file)
  if (category) {
    formData.append('category', category)
  }
  return await axios.post('/words/upload-wordbook', formData)
}
